sky/skylet/constants.py

```diff
@@ -66,19 +74,38 @@
     DOCKER_SERVER_ENV_VAR,
 }
 
+ACTIVATE_PYTHON_ENV = (f'[ -d {SKY_REMOTE_PYTHON_ENV} ] && '
+                       f'source {SKY_REMOTE_PYTHON_ENV}/bin/activate;')
+
+
+def run_in_python_env(command):
+    """Returns a command that runs in the SkyPilot's python environment."""
+    return f'( {ACTIVATE_PYTHON_ENV}{command}; deactivate )'
+
+
 # Install conda on the remote cluster if it is not already installed.
 # We use conda with python 3.10 to be consistent across multiple clouds with
 # best effort.
 # https://github.com/ray-project/ray/issues/31606
 # We use python 3.10 to be consistent with the python version of the
 # AWS's Deep Learning AMI's default conda environment.
+_RUN_PYTHON = run_in_python_env('python \\$@')
+_RUN_PIP = run_in_python_env('pip \\$@')
+_RUN_RAY = run_in_python_env('ray \\$@')
 CONDA_INSTALLATION_COMMANDS = (
     'which conda > /dev/null 2>&1 || '
-    '(wget -nc https://repo.anaconda.com/miniconda/Miniconda3-py310_23.11.0-2-Linux-x86_64.sh -O Miniconda3-Linux-x86_64.sh && '  # pylint: disable=line-too-long
+    '{ wget -nc https://repo.anaconda.com/miniconda/Miniconda3-py310_23.11.0-2-Linux-x86_64.sh -O Miniconda3-Linux-x86_64.sh && '  # pylint: disable=line-too-long
     'bash Miniconda3-Linux-x86_64.sh -b && '
     'eval "$(~/miniconda3/bin/conda shell.bash hook)" && conda init && '
-    'conda config --set auto_activate_base true); '
-    'grep "# >>> conda initialize >>>" ~/.bashrc || conda init;')
+    'conda config --set auto_activate_base true && source ~/.bashrc; }; '
+    'grep "# >>> conda initialize >>>" ~/.bashrc || conda init;'
+    # Create a separate conda environment for SkyPilot dependencies.
+    f'[ -d {SKY_REMOTE_PYTHON_ENV} ] || '
+    f'python -m venv {SKY_REMOTE_PYTHON_ENV}; '
```

Comment:
Have we considered using a conda env instead? What are the pros & cons of the two implementations?